### PR TITLE
[mongo-c-driver,libbson] Update and fixes

### PR DESCRIPTION
**Pull request by dg0yt**
- #### What does your PR fix?
  Updates mongo-c-driver and its split-out libbson.
  Major cleanup.
  Fixes exported pkg-config data, needed for #23898.

- #### Which triplets are supported/not supported? Have you updated the [CI baseline](https://github.com/microsoft/vcpkg/blob/master/scripts/ci.baseline.txt)?
  libbson: all, mongo-c-driver: !uwp; yes

- #### Does your PR follow the [maintainer guide](https://github.com/microsoft/vcpkg/blob/master/docs/maintainers/maintainer-guide.md)?
  yes

- #### If you have added/updated a port: Have you run `./vcpkg x-add-version --all` and committed the result?
  yes

### CODE REVIEW

File: ports/mongo-c-driver/vcpkg.json [hunk 1]
```diff
@@ -27,6 +31,12 @@
         "icu"
       ]
     },
+    "openssl": {
+      "description": "Use OpenSSL (even on Window or Apple systems)",
```
Comment (Thomas1664) [inline ports/mongo-c-driver/vcpkg.json:35]: Veeery nitpicky but Windows is missing an 's' here...

**Review by JonLiu1993 — DISMISSED** [review]

File: ports/mongo-c-driver/vcpkg.json [hunk 2]
```diff
@@ -1,14 +1,18 @@
 {
   "name": "mongo-c-driver",
-  "version": "1.17.6",
-  "port-version": 2,
+  "version": "1.22.2",
   "description": "Client library written in C for MongoDB.",
   "homepage": "https://github.com/mongodb/mongo-c-driver",
+  "license": null,
```
Comment (JonLiu1993) [inline ports/mongo-c-driver/vcpkg.json:6]: @dg0yt , Thanks for your pr, Isn't `Apache-2.0 license` the license of mongo-c-driver?
https://github.com/mongodb/mongo-c-driver/blob/master/COPYING

Comment (dg0yt) [reply ports/mongo-c-driver/vcpkg.json:6]: That's the main license, but some parts use other licenses. Note the file list with `vcpkg_install_copyright`. Or do what `null` signals: read the installed `copyright`.

**Comment by JonLiu1993** [comment]
@dg0yt , When I tested the features by command `./vcpkg install mongo-c-driver[*]` I got this error.
Please take a look:
````
"optimized;F:/Feature-test/mongo-c-driver/vcpkg/installed/x86-windows/lib/zlib.lib;debug;F:/Feature-test/mongo-c-driver/vcpkg/installed/x86-windows/debug/lib/zlibd.lib"
-- Looking for include file unistd.h
-- Looking for include file unistd.h - not found
-- Looking for include file stdarg.h
-- Looking for include file stdarg.h - found
-- Searching for compression library zstd
-- Could NOT find PkgConfig (missing: PKG_CONFIG_EXECUTABLE) 
CMake Error at src/libmongoc/CMakeLists.txt:122 (message):
    Not found


-- Configuring incomplete, errors occurred!
See also "F:/Feature-test/mongo-c-driver/vcpkg/buildtrees/mongo-c-driver/x86-windows-dbg/CMakeFiles/CMakeOutput.log".
See also "F:/Feature-test/mongo-c-driver/vcpkg/buildtrees/mongo-c-driver/x86-windows-dbg/CMakeFiles/CMakeError.log".
````

**Comment by dg0yt** [comment]
> ~~~
> -- Searching for compression library zstd
> -- Could NOT find PkgConfig (missing: PKG_CONFIG_EXECUTABLE

I don't think it is a regression. Untested fix pushed nevertheless.

**Comment by JonLiu1993** [comment]
> > ```
> > -- Searching for compression library zstd
> > -- Could NOT find PkgConfig (missing: PKG_CONFIG_EXECUTABLE
> > ```
> 
> I don't think it is a regression. Untested fix pushed nevertheless.

@dg0yt , This error persists after I retest.

**Comment by dg0yt** [comment]
Host pkgconf needs host triplet. Resolved.

**Comment by JonLiu1993** [comment]
All features are tested successfully in the following triplet:
- x86-windows
- x64-windows
- x64-windows-static

**Review by JonLiu1993 — APPROVED** [review]

**Comment by BillyONeal** [comment]
I'm sorry I didn't get to this one today, just letting you know I didn't miss it, I just ran out of time

File: ports/mongo-c-driver/fix-mingw.patch [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+ 
+diff --git a/src/libmongoc/src/mongoc/mongoc-sspi-private.h b/src/libmongoc/src/mongoc/mongoc-sspi-private.h
+index 60996fc..3745eff 100644
+--- a/src/libmongoc/src/mongoc/mongoc-sspi-private.h
++++ b/src/libmongoc/src/mongoc/mongoc-sspi-private.h
+@@ -28,7 +28,7 @@ BSON_BEGIN_DECLS
+ 
+ #define SECURITY_WIN32 1 /* Required for SSPI */
+ 
+-#include <Windows.h>
```
Comment (BillyONeal) [inline ports/mongo-c-driver/fix-mingw.patch:72]: FWIW, this change seems backwards: the canonical name of this header is capitalized:

image

I'm deferring to you on what MinGW should do but not sure this is correct.

File: ports/mongo-c-driver/fix-mingw.patch [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+diff --git a/src/libmongoc/CMakeLists.txt b/src/libmongoc/CMakeLists.txt
+index 2e1f854..95e6b6b 100644
+--- a/src/libmongoc/CMakeLists.txt
++++ b/src/libmongoc/CMakeLists.txt
+@@ -688,7 +688,7 @@ if (NOT ENABLE_SSL STREQUAL OFF)
+          ${PROJECT_SOURCE_DIR}/src/mongoc/mongoc-stream-tls-secure-channel.c
+          ${PROJECT_SOURCE_DIR}/src/mongoc/mongoc-secure-channel.c
+       )
+-      set (SSL_LIBRARIES secur32.lib crypt32.lib Bcrypt.lib)
++      set (SSL_LIBRARIES secur32.lib crypt32.lib bcrypt.lib)
```
Comment (BillyONeal) [inline ports/mongo-c-driver/fix-mingw.patch:23]: This fixes it to match the canonical name. No change requested.

File: ports/mongo-c-driver/fix-mingw.patch [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+    elseif (LIBRESSL)
+       message (STATUS "Compiling against LibreSSL")
+       set (SOURCES ${SOURCES}
+@@ -714,7 +714,7 @@ if (MONGOC_ENABLE_SASL)
+       message (STATUS "Compiling against Windows SSPI")
+       set (SOURCES ${SOURCES} ${PROJECT_SOURCE_DIR}/src/mongoc/mongoc-cluster-sspi.c)
+       set (SOURCES ${SOURCES} ${PROJECT_SOURCE_DIR}/src/mongoc/mongoc-sspi.c)
+-      set (SASL_LIBRARIES secur32.lib crypt32.lib Shlwapi.lib)
++      set (SASL_LIBRARIES secur32.lib crypt32.lib shlwapi.lib)
```
Comment (BillyONeal) [inline ports/mongo-c-driver/fix-mingw.patch:32]: image

OK the canonical name for this one seems crazy O_O

File: ports/mongo-c-driver/fix-mingw.patch [hunk 6]
```diff
@@ -0,0 +1,76 @@
+diff --git a/build/cmake/FindResSearch.cmake b/build/cmake/FindResSearch.cmake
+index fcd791b..346f715 100644
+--- a/build/cmake/FindResSearch.cmake
++++ b/build/cmake/FindResSearch.cmake
+@@ -2,7 +2,7 @@ include (CheckSymbolExists)
+ 
+ if (ENABLE_SRV STREQUAL ON OR ENABLE_SRV STREQUAL AUTO)
+    if (WIN32)
+-      set (RESOLV_LIBRARIES Dnsapi)
++      set (RESOLV_LIBRARIES dnsapi)
```
Comment (BillyONeal) [inline ports/mongo-c-driver/fix-mingw.patch:10]: The canonical name of this one is `DnsAPI`
image

**Review by BillyONeal — APPROVED** [review]
Thanks!

File: ports/mongo-c-driver/fix-mingw.patch (hunk 6 above)
Comment (dg0yt) [reply ports/mongo-c-driver/fix-mingw.patch:10]: I don't how it was ten years ago, and how it was documented ten years ago. But I know the spelling in ubuntu mingw packages, and the filesystem is case-sensitive on Linux. It really makes a difference.

Comment (BillyONeal) [reply ports/mongo-c-driver/fix-mingw.patch:10]: The problem is that, specifically for compatibility with \*nix, there are compiler warnings that fire if one gets this case wrong. The canonical owner of these things is the Platform SDK, so MinGW should be matching that.

The oldest Windows SDK to which I have access had the same case, at least for `Windows.h`, and I can't imagine they would have changed any of these.

**Pull request merged**